0000-lossy-conversions.md

```diff
@@ -0,0 +1,310 @@
+
+   I do not think that this is a deal-breaker. Rustaceans have come to accept that you need `.unwrap()` to access an optional value, and `Box::new()` to allocate heap memory: things that many popular languages do automatically. But since Rust had a more concise way of converting integers, and may now abandon it, people might be unhappy because they will have to change their coding habits. Furthermore, the new way isn't _obviously_ better than the old one in every way. Probably only those who have had to deal with integer truncation bugs will fully appreciate this change.
+
+2. The API surface of this change is rather big: Most numeric types get 3 new methods.
+
+3. This may change the overall character of the language.
+
+   However, I believe it would make the language feel more consistent, since Rust already leans towards explicitness in most other situations.
+
+4. This may negatively impact compile times _(to be verified)_.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+First, I'd like to compare this with integer arithmetic. An expression like `x + 200_u8` can overflow, which is implicit, similarly to integer casts. However, there are some crucial differences:
+
+- Arithmetic is more common than conversions, so there is a bigger need to be concise
+- For arithmetic, there already exist checked, truncating and wrapping methods. For example, you can write `200_u8.saturating_add(x)` if you want; an equivalent for conversions does not exist.
+- By default, overflow on arithmetic operations wraps around in release builds and panics in debug builds. This means that bugs due to integer overflow can be caught with tests. Integer conversions on the other hand are _always_ unchecked. This means that bugs due to truncation are easier to miss.
+
+Therefore, I believe that making lossy conversions more explicit would go a long way toward avoiding bugs, making you more confident that your code is correct, and saving you time debugging and writing tests. But is the proposal outlined above the best possible solution? Let's look at a few alternatives:
+
+1. We could add the conversion methods and the `cast_lossy` lint, but never turn it into an error. It would remain a warning, which people are free to ignore or disable. This makes sense if avoiding the `as` operator is seen more as a stylistic preference than a correctness issue.
+
+2. `saturate()` isn't needed to deprecate `as` for lossy numeric conversions, so we could add only `truncate()` and `approx()`.
+
+   However, the standard library already contains saturating math operations, so adding saturating conversions makes sense.
+
+3. Instead of deprecating `as` only for lossy numeric casts, it could be deprecated for all numeric casts, so `From`/`Into` is required in these situations.
+
+   This feels like overkill. If people really want to forbid `as` for lossless conversions, they can use clippy's `cast_lossless` lint.
```

Comment:
Is there any justification that people do not already use `cast_lossless`? Or might it simply be because it is too troublesome to do `.into()` *and* specify the proper type (because we don't have `.into::<T>()`)?

Reply:
Currently, `cast_lossless` isn't used because of the more verbose syntax. The [rationale](https://rust-lang.github.io/rust-clippy/master/#cast_lossless) for this lint is: `as` can sometimes perform lossy conversions, so using it even for lossless conversions is bad, because it may silently become lossy when changing types while refactoring.

Once lossy `as` casts produce a warning, there will be no benefit to forbidding lossless `as` casts.